In [1]:
from urllib.request import urlretrieve
import pandas as pd

In [2]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url, 'd') # Trying to retrive the csv file attachted to the url

('d', <http.client.HTTPMessage at 0x23874be3110>)

In [3]:
covid_df=pd.read_csv('italy-covid-daywise.csv') # Pandas reading csv file

In [4]:
type(covid_df) # Tying to check the type of the data

pandas.core.frame.DataFrame

In [5]:
covid_df  # Printing the values stored by pandas

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

we can view some basic information about the data frame using the **.info** method

covid_df.info()

It appears that each column contians a values of a specific data type. For the numeric columns, you can view
some statistical information like mean, standard deviation, minimum/maximum values and the number of non-empty
values using the **.describe** method

In [6]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

The columns properlycontains the list of the columns with the data frame

In [7]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

You can also retrieve the number of row and columns in data frame using the **.shape** method

In [8]:
covid_df.shape

(248, 4)

In [9]:
covid_df.at[246,'new_cases']

np.float64(975.0)

You're seeing **np.float64(975.0)** even though you didn't import numpy because Pandas internally uses NumPy,
and some of its operations (like data access with .at[], .iloc[], etc.) may return NumPy data types such as np.float64.

So if you wish to work with native python float then use:
**float(covid_df.at[246, 'new_cases'])**

## Why NumPy is used (especially by Pandas):
NumPy (short for Numerical Python) is a powerful library that provides:

* Efficient multidimensional arrays (ndarray)
* Fast mathematical operations on these arrays
* Built-in support for linear algebra, statistics, random sampling, etc.

In short Pandas depends on NumPy for:
* Speed
* Memory Effiency
* Mathematical Funtionality
* Data Type Consistancy

In [10]:
float(covid_df.at[246, 'new_cases'])

975.0

Instead of using indexing notation [], Pandas also allows accessing columns as properties of the data frame using **"." (dot)** notation.
However this method only works for column whose name do not contain spaces or special characters.

In [11]:
covid_df.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

you can also pass a list of columns within the indexing notation **[]** to access the subset of the data frame with just the
given columns

In [12]:
cases_df = covid_df[['date', 'new_cases']]

In [13]:
cases_df

date  new_cases
0    2019-12-31        0.0
1    2020-01-01        0.0
2    2020-01-02        0.0
3    2020-01-03        0.0
4    2020-01-04        0.0
..          ...        ...
243  2020-08-30     1444.0
244  2020-08-31     1365.0
245  2020-09-01      996.0
246  2020-09-02      975.0
247  2020-09-03     1326.0

[248 rows x 2 columns]

In order to access a specific row of data use **.loc** method

In [14]:
covid_df.loc[243]

date          2020-08-30
new_cases         1444.0
new_deaths           1.0
new_tests        53541.0
Name: 243, dtype: object

In [15]:
type(covid_df.loc[243])

pandas.core.series.Series

to find the first and the last row of data, we can use the **.head()** and **.tail()** methods.

In [16]:
covid_df.head(5)  # To get the first 5 rows of data

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN
3  2020-01-03        0.0         0.0        NaN
4  2020-01-04        0.0         0.0        NaN

In [17]:
covid_df.tail(4)   # To get the last 4 rows of data

date  new_cases  new_deaths  new_tests
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

The distinction between *0* and *NaN* is subtle but important. In this data set it represents that daily tests numbers
were not reported on specific dates. In fact Italy started reportingdaily test from April 19, 2020 by thatt time, 935310
tests had already been conducted.

We can find the first index which doesn't contain *NaN* value by using **first_valid_index()** method of a series.

In [18]:
covid_df.new_tests.first_valid_index()

111

To check if the value of the new tests is changing from NaN value to an actual floating number then **use a range function in the .loc[]** method

In [19]:
covid_df.loc[108 : 113]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0

Use **.sample()** method to retirieve a random sample of rows from the data frame

In [20]:
covid_df.sample(10)

date  new_cases  new_deaths  new_tests
148  2020-05-27      397.0        78.0    37299.0
187  2020-07-05      235.0        21.0    21166.0
101  2020-04-10     4204.0       612.0        NaN
171  2020-06-19      331.0        66.0    28570.0
92   2020-04-01     4053.0       839.0        NaN
27   2020-01-27        0.0         0.0        NaN
95   2020-04-04     4585.0       764.0        NaN
200  2020-07-18      231.0        11.0    27569.0
80   2020-03-20     5322.0       429.0        NaN
120  2020-04-29     2091.0       382.0    38589.0

# Analyzing Data from the data frames

#### Q1: What is the total number of reported cases and deaths to covi-19 in Italy?
-> *Similar to NumPy arrays, a Pandas series supports* **sum** *method to answer these question*

In [21]:
total_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()

In [22]:
print(f"The total number of new cases in Italy during covid-19 is: {int(total_cases)} \nThe total number of deaths in Italy during covid-19 is: {int(total_deaths)}")

The total number of new cases in Italy during covid-19 is: 271515 
The total number of deaths in Italy during covid-19 is: 35497


#### Q2: What is the overal death rate in Italy during covid-19?

In [23]:
death_rate = total_deaths / total_cases

In [24]:
print(f"The overal death rate in Italy during the covid-19 is: {death_rate:.4f}%")

The overal death rate in Italy during the covid-19 is: 0.1307%


#### Q3: What is the overal number of tests conducted if a total of 935310 tests were conducted before the daily tests were being reported?

In [25]:
total_tests = 935310 + covid_df.new_tests.sum()

In [26]:
print(f"The total number of test which were conducted in Italy during the covid-19 is: {int(total_tests)}")

The total number of test which were conducted in Italy during the covid-19 is: 5214766


#### Q4: What fraction of the test was returned a positive result?

In [27]:
positive_tests = total_cases / covid_df.new_tests.sum()

In [28]:
print(f"The total number of tests which turned out to be positive are: {positive_tests:.4f}%")

The total number of tests which turned out to be positive are: 0.0634%


#### Q5: Print all the rows of data that have 1,000 or more cases.

In [29]:
high_new_cases = covid_df.new_cases >= 1000

In [30]:
high_new_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

In [31]:
high_cases = covid_df[high_new_cases]

In [32]:
print(f"The rows of data which had 1000 or more new cases in a day are: \n{high_cases}")

The rows of data which had 1000 or more new cases in a day are: 
           date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]


Well we colud also reduce them into one single line

Like:

In [33]:
print(f"The rows of data which had 1000 or more cases in a day are:\n{covid_df[covid_df.new_cases >= 1000]}")

The rows of data which had 1000 or more cases in a day are:
           date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]


The data frame contains 72 rows but only first 5 and the last 5 rows of data are being displayed by deafult
jupyter, for brevity. So if you wish to go through all the rows, then we can modify some display options.

In [34]:
#from IPython.display import display
#with pd.option_context('display.max_rows',100):   # The 100 here limits the DataFrame display to 100 rows within this block
#    display(covid_df[covid_df.new_cases >= 1000])

## How this works:

**1. from IPython.display import diaplay**'

*This imports the display() function, which is used in Jupyter Notebooks 
(and IPython environments) to nicely format and show objects like DataFrames, HTML, etc.*

**2. with pd.option_context('display.max_rows', 100):**

*That means:
Inside this with block, Pandas will show up to 100 rows when displaying a DataFrame.
Once the block is done, Pandas will return to its previous setting.*

**3. covid_df[covid_df.new_cases >= 1000]**

*This is a boolean filter on the DataFrame covid_df. It selects only the rows where the 
new_cases column has values greater than or equal to 1000.*
*Which will return a subset of the original DataFrame that matches this condition.*

#### Q6: Determine the days when the ratio of cases reported to test conducted is higher that the overal **positive_rate**

In [35]:
positive_tests

np.float64(0.06344614829548428)

In [36]:
ratio_df = covid_df[(covid_df.new_cases / covid_df.new_tests) > positive_tests]

In [37]:
print(f"The days when the ratio of cases reported to test conducted is higher than the overal Positive rate is: \n{ratio_df}")

The days when the ratio of cases reported to test conducted is higher than the overal Positive rate is: 
           date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
124  2020-05-03     1900.0       474.0    27047.0
128  2020-05-07     1444.0       369.0    13665.0


In [38]:
covid_df.new_cases / covid_df.new_tests

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
243    0.026970
244    0.032055
245    0.018311
246         NaN
247         NaN
Length: 248, dtype: float64

Well we could also use this series to add a new column to the data frame

In [39]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests

In [40]:
covid_df

date  new_cases  new_deaths  new_tests  positive_rate
0    2019-12-31        0.0         0.0        NaN            NaN
1    2020-01-01        0.0         0.0        NaN            NaN
2    2020-01-02        0.0         0.0        NaN            NaN
3    2020-01-03        0.0         0.0        NaN            NaN
4    2020-01-04        0.0         0.0        NaN            NaN
..          ...        ...         ...        ...            ...
243  2020-08-30     1444.0         1.0    53541.0       0.026970
244  2020-08-31     1365.0         4.0    42583.0       0.032055
245  2020-09-01      996.0         6.0    54395.0       0.018311
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

[248 rows x 5 columns]

## Points to be noted:

Sometimes it takes a few days to get the results of a test, so we can't really compare the number of new cases with the number of tests conducted on the same day. Any inference based on ***this positive_rate is most likely to be incorrect***. It's important to watch out for subtle relationships like these, which are often not conveyed within the CSV file and require some external context. It's always a good idea to read through the documentation provided with the dataset and ask for more information.

So in order to remove the **positive_rate** column we will use the ***drop*** method.

In [41]:
covid_df.drop(columns = ['positive_rate'], inplace = True)

In [42]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

## Purpose of inplace=True:

The inplace=True argument tells Pandas to modify the DataFrame directly rather than returning a new modified copy.

#### If we don't use inplace:

***covid_df = covid_df.drop(columns=['positive_rate'])***

* The method returns a new DataFrame without the **'positive_rate'** column.

* You must assign the result back to covid_df to save the change.

#### When we use inplace:

***covid_df.drop(columns=['positive_rate'], inplace=True)***

* The original covid_df is changed in-place, and the **'positive_rate'** column is removed directly.

* It returns None, so no reassignment is needed.

### Sorting rows with the column values

The roows can also be sorted by a specific column using ***.sort_values***.

Let's sort to identify the dayse which had the highest number of cases, then chain it up with the ***head()*** method to get the 
10 days with the most cases

In [43]:
covid_df.sort_values('new_cases', ascending = False).head(10)

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN
83  2020-03-23     5560.0       649.0        NaN
80  2020-03-20     5322.0       429.0        NaN
85  2020-03-25     5249.0       743.0        NaN
90  2020-03-30     5217.0       758.0        NaN
86  2020-03-26     5210.0       685.0        NaN

It looks like the last week of March had the highest number of daily cases. Let's compare this with the days
when the highest number of deaths were recorded.

In [44]:
covid_df.sort_values('new_deaths',ascending = False).head(10)

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN
95  2020-04-04     4585.0       764.0        NaN
94  2020-04-03     4668.0       760.0        NaN
90  2020-03-30     5217.0       758.0        NaN
85  2020-03-25     5249.0       743.0        NaN
93  2020-04-02     4782.0       727.0        NaN

It appears that daily deaths peaked about a week after the peak in daily new cases.

Let’s also look at the days with the lowest number of cases.

In [45]:
covid_df.sort_values('new_cases',ascending = True).head(10)

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
0    2019-12-31        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
5    2020-01-05        0.0         0.0        NaN
6    2020-01-06        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
8    2020-01-08        0.0         0.0        NaN
9    2020-01-09        0.0         0.0        NaN

It looks like the count of new cases on June 20th was -148 — a negative number, which we didn’t expect to see. But that’s the nature of real-world data. It could be a data entry error, or perhaps the government issued a correction for a previous miscount.

If this was indeed a data entry error, then we can use one of the following approaches to deal with the missing or faulty values:

1. Replace it with **0**

2. Replace it with the average of the entire column

3. Replace it with the average of the values on the previous and the next date

4. Discard the row entirely

Whichever approach you choose requires some context about the data and the problem. In this case, since the data is ordered by dates, we can go with the third approach.

The **.at** method can be used to modify a specific value within the DataFrame.

In [46]:
covid_df.loc[170 : 175]

date  new_cases  new_deaths  new_tests
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20     -148.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

In [47]:
covid_df.at[172,'new_cases'] = (covid_df.at[171, 'new_cases'] + covid_df.at[173, 'new_cases']) * 0.5

In [48]:
covid_df.loc[170 : 175]

date  new_cases  new_deaths  new_tests
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20      297.5        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

## Best Practices for Managing Data Corrections and Anomalies
The government corrects a past mistake using a negative value, and I overwrite it with a calculated mean (like the average of neighboring values), am I removing a real correction and making the data less accurate?

***Yes — potentially.***

If that negative value was intentional and meaningful, then replacing it blindly could distort the truth. You're losing the correction and inserting your own assumption.

And so that's why we should know about the context of the data first.

Here are some data aware approaches:
1. Understand the context first-
   * Was the negative value an error, or a correction?
   * Did the source mention a data revision or reclassification?

2. Don’t rush to clean — annotate instead:
   * You can keep the negative value, but flag it in a new column like note = "government correction".

3. Create two different versions:
   * Keep one dataset original, and make a cleaned copy only for analysis that needs it.
   * You can then test how much your imputation changes the trends.

4. Be transparent:
   * Always mention in your report or code comments: “Value modified due to government correction; original value was -148.”


Many governments during COVID-19 revised counts retroactively:

* Some days showed negative new cases due to removal of duplicates.
* Others showed spikes from delayed reporting.

Analysts would:

* Keep the original data intact for auditability
* But smooth the data when plotting trends, using rolling averages or filtered views

In [49]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

## Working with dates
The data type of dates is currently an object, so Pandas doesn't recognize that this column contains dates.
However, we can convert it into a datetime column using the ***pd.to_datetime*** method.

In [50]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [51]:
covid_df.date

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

Now, as you can see, the data type of **covid_df['date']** has changed from object to datetime64[ns] (where ns indicates that the 
date and time are stored as float64 values with nanosecond precision).
We can now extract different parts of the date into separate columns using ***DatetimeIndex*** class.

In [52]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [53]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5
..         ...        ...         ...        ...   ...    ...  ...      ...
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3

[248 rows x 8 columns]

If in case you are wondering about the 0 1 2 3 4 5 6 and so on the weekday column it's all because 
the numbers in the weekday column represent the day of the week, and they are based on
Python convention where:
* 0 = Monday
* 1 = Tuesday
* 2 = Wednusday
* 3 = Thrusday
* 4 = Friday
* 5 = Saturday
* 6 = Sunday

So in the data:

**2019-12-31 → weekday = 1 → Tuesday  
2020-01-01 → weekday = 2 → Wednesday  
2020-01-02 → weekday = 3 → Thursday  
2020-01-03 → weekday = 4 → Friday  
2020-01-04 → weekday = 5 → Saturday**

If you wish to add the weekday names (like "Tuesday", "Wednesday"), you can create a new column like this:

First drop the existing coulmn with the numerical weekday:

***covid_df.drop(columns = ['weekday'], inplace = True)***

Then add the new weekday which contains the week_name

***covid_df['weekday_name'] = covid_df['date'].dt.day_name()***

The same could be done for the months as well. 
To get the month names (like January, February, etc.) from a datetime column in Pandas,
you can use the ***.dt.month_name()*** method — just like you did with weekday names.

***covid_df['month_name'] = covid_df['date'].dt.month_name()***

In [54]:
covid_df.drop(columns = ['weekday'], inplace = True)

In [55]:
covid_df.drop(columns = ['month'], inplace = True)

In [56]:
covid_df

date  new_cases  new_deaths  new_tests  year  day
0   2019-12-31        0.0         0.0        NaN  2019   31
1   2020-01-01        0.0         0.0        NaN  2020    1
2   2020-01-02        0.0         0.0        NaN  2020    2
3   2020-01-03        0.0         0.0        NaN  2020    3
4   2020-01-04        0.0         0.0        NaN  2020    4
..         ...        ...         ...        ...   ...  ...
243 2020-08-30     1444.0         1.0    53541.0  2020   30
244 2020-08-31     1365.0         4.0    42583.0  2020   31
245 2020-09-01      996.0         6.0    54395.0  2020    1
246 2020-09-02      975.0         8.0        NaN  2020    2
247 2020-09-03     1326.0         6.0        NaN  2020    3

[248 rows x 6 columns]

In [57]:
covid_df['weekday_name'] = covid_df['date'].dt.day_name()

In [58]:
covid_df['month_name'] = covid_df['date'].dt.month_name()

In [59]:
covid_df

date  new_cases  new_deaths  new_tests  year  day weekday_name  \
0   2019-12-31        0.0         0.0        NaN  2019   31      Tuesday   
1   2020-01-01        0.0         0.0        NaN  2020    1    Wednesday   
2   2020-01-02        0.0         0.0        NaN  2020    2     Thursday   
3   2020-01-03        0.0         0.0        NaN  2020    3       Friday   
4   2020-01-04        0.0         0.0        NaN  2020    4     Saturday   
..         ...        ...         ...        ...   ...  ...          ...   
243 2020-08-30     1444.0         1.0    53541.0  2020   30       Sunday   
244 2020-08-31     1365.0         4.0    42583.0  2020   31       Monday   
245 2020-09-01      996.0         6.0    54395.0  2020    1      Tuesday   
246 2020-09-02      975.0         8.0        NaN  2020    2    Wednesday   
247 2020-09-03     1326.0         6.0        NaN  2020    3     Thursday   

    month_name  
0     December  
1      January  
2      January  
3      January  
4      January  
..         ...  
243     August  
244     August  
245  September  
246  September  
247  September  

[248 rows x 8 columns]

Let's check the overal matrics for the month of May. We can query the row for May, choose a subset that we want to aggregate,
and use the **sum** method of the Data Fame to get the sum of the values of each chosen columns.

In [60]:
# Query for the row for May
covid_df_may = covid_df[covid_df['month_name'] == 'May']

# Extracting the subset of the column to be aggregated
covid_df_may_matric = covid_df_may[['new_cases', 'new_deaths', 'new_tests']]

# Now getting the column-wise sum
covid_may_total = covid_df_may_matric.sum()

In [61]:
covid_may_total

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

In [62]:
type(covid_may_total)

pandas.core.series.Series

Well apperantly the above operations can be performed within a single line.

In [63]:
covid_df_may_total = covid_df[covid_df['month_name'] == 'May'][['new_cases', 'new_deaths', 'new_tests']].sum()

In [64]:
covid_df_may_total

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

Let's check if the number of cases reported on Sunday is higher than the average number of cases reported every day.

But this time we will use the **.mean()** method to aggregate

In [65]:
# Overal average 
covid_avg = covid_df.new_cases.mean()

In [66]:
covid_avg_may = covid_df[covid_df['weekday_name'] == 'Sunday'].new_cases.mean()

In [67]:
if covid_avg > covid_avg_may:
    print("The average number of cases in the entire data frame is higher than that of Sundays.")
else:
    print("The average number of cases reported on Sundays was actually higher than the overall average in the data frame.")

The average number of cases reported on Sundays was actually higher than the overall average in the data frame.


## Grouping and Aggregation
We might want to summarize the day-wise DataFrame and create a new DataFrame with month-wise data. This is where the **groupby** function becomes useful. Along with grouping, we also need to specify a method to aggregate the data within each group.

In [68]:
covid_month_df = covid_df.groupby('month_name')[['new_cases', 'new_tests', 'new_deaths']].sum()

In [69]:
covid_month_df

new_cases  new_tests  new_deaths
month_name                                  
April        101852.0   419591.0     16091.0
August        21060.0  1098704.0       345.0
December          0.0        0.0         0.0
February        885.0        0.0        21.0
January           3.0        0.0         0.0
July           6722.0   797692.0       388.0
June           8217.5   830354.0      1404.0
March        100851.0        0.0     11570.0
May           29073.0  1078720.0      5658.0
September      3297.0    54395.0        20.0

The above table seems a bit unordered, so if you wish to assign it a specific order, you could use:

Here we are giving the pandas the order of values we want in our DataFrame

In [70]:
# Defining the correct order of months
month_order = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']

In [71]:
# Making sure 'month_name' column exists and is ordered correctly
covid_df['month_name'] = pd.Categorical(covid_df['month_name'], categories=month_order, ordered=True)

Well the above codes can be described as

covid_df['month_name'] = pd.Categorical(

    covid_df['month_name'],           # the column to convert
    
    categories=month_order,           # the correct month order
    
    ordered=True                      # tells pandas to treat it as ordered
)

Now when you do:

**covid_month_df = covid_df.groupby('month_name')[['new_cases', 'new_tests', 'new_deaths']].sum().sort_index()**

It ussees the calender order you defined previously—instead of defaulting to A–Z.

In [72]:
# Now group and sort
covid_month_df = covid_df.groupby('month_name')[['new_cases', 'new_tests', 'new_deaths']].sum().sort_index()

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_20072\1730805980.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  covid_month_df = covid_df.groupby('month_name')[['new_cases', 'new_tests', 'new_deaths']].sum().sort_index()


## Brainstorming

### What is pd.Categorical?
In Pandas, a categorical variable is like a column that contains fixed possible values, often used for:

* Sorting

* Grouping

* Reducing memory

By default, strings like "April", "May", etc., are treated as plain text and are sorted alphabetically, not in calendar order.

### Why Categorical Helps??
When you convert a column (like month_name) into a Categorical,so what you doing is you’re telling Pandas:

**“Hey, here's the specific order of these values — please follow this order when sorting, grouping, or displaying them.”**

### Future Error??
Well, the output seems promising, but why does this show up?

**FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.**

**covid_month_df = covid_df.groupby('month_name')[['new_cases', 'new_tests', 'new_deaths']].sum().sort_index()**

This warning is displayed by Pandas. It's basically saying:

**“Hey, when you're grouping by a categorical column like month_name, we used to include all categories — even if some of them don’t appear in your data.
But soon, the default behavior will change to only include the ones that are actually present.”**

This happens because:

* You converted month_name to a Categorical with all 12 months.

* But your data might only have, say, 10 months (there is no October, November and December in our DataFrame).

* Pandas includes all 12 anyway (because observed=False is the current default).

* This will change in future Pandas versions — so they warn you now.

In [73]:
covid_month_df

new_cases  new_tests  new_deaths
month_name                                  
January           3.0        0.0         0.0
February        885.0        0.0        21.0
March        100851.0        0.0     11570.0
April        101852.0   419591.0     16091.0
May           29073.0  1078720.0      5658.0
June           8217.5   830354.0      1404.0
July           6722.0   797692.0       388.0
August        21060.0  1098704.0       345.0
September      3297.0    54395.0        20.0
October           0.0        0.0         0.0
November          0.0        0.0         0.0
December          0.0        0.0         0.0

Apart from grouping, there is another form of aggregation which is used to calculate running or the cumulative sum of cases, tests, or deaths upto the current
date for each row. This could be using **cumsum** method. Let's add three new columns: **total_cases**, **total_deaths** and **total_tests**

In [74]:
covid_df['total_cases'] = covid_df.new_cases.cumsum()

In [75]:
covid_df['total_deaths'] = covid_df.new_deaths.cumsum()

In [76]:
covid_df['total_tests'] = covid_df.new_tests.cumsum() 

In [77]:
covid_df

date  new_cases  new_deaths  new_tests  year  day weekday_name  \
0   2019-12-31        0.0         0.0        NaN  2019   31      Tuesday   
1   2020-01-01        0.0         0.0        NaN  2020    1    Wednesday   
2   2020-01-02        0.0         0.0        NaN  2020    2     Thursday   
3   2020-01-03        0.0         0.0        NaN  2020    3       Friday   
4   2020-01-04        0.0         0.0        NaN  2020    4     Saturday   
..         ...        ...         ...        ...   ...  ...          ...   
243 2020-08-30     1444.0         1.0    53541.0  2020   30       Sunday   
244 2020-08-31     1365.0         4.0    42583.0  2020   31       Monday   
245 2020-09-01      996.0         6.0    54395.0  2020    1      Tuesday   
246 2020-09-02      975.0         8.0        NaN  2020    2    Wednesday   
247 2020-09-03     1326.0         6.0        NaN  2020    3     Thursday   

    month_name  total_cases  total_deaths  total_tests  
0     December          0.0           0.0          NaN  
1      January          0.0           0.0          NaN  
2      January          0.0           0.0          NaN  
3      January          0.0           0.0          NaN  
4      January          0.0           0.0          NaN  
..         ...          ...           ...          ...  
243     August     267298.5       35473.0    4182478.0  
244     August     268663.5       35477.0    4225061.0  
245  September     269659.5       35483.0    4279456.0  
246  September     270634.5       35491.0          NaN  
247  September     271960.5       35497.0          NaN  

[248 rows x 11 columns]

In [78]:
covid_df.total_tests.tail(10)

238    3883814.0
239    3941868.0
240    3999508.0
241    4064643.0
242    4128937.0
243    4182478.0
244    4225061.0
245    4279456.0
246          NaN
247          NaN
Name: total_tests, dtype: float64

## Merging data drom multiple sources